### ISSUE TITLE
Post-merge review of #70: phpstan analyse is red again, duplicate suppression is mostly ineffective, trait deferral is partial

### ISSUE THREAD

**Issue by voku** [voku/phpstan-rules#79]
Review of `master` at 1a66c78 (PRs #71–#78), against the base 045225e.

First the good news: the merged suite is green — 122 tests, 716 assertions, 2 skipped, on `phpstan/phpstan` 2.2.9. All four blockers raised during review of #70 are genuinely addressed: the `IfConditionMatchRule` / `IfConditionSwitchCaseRule` constructor positions are preserved (no BC break), `%treatPhpDocTypesAsCertain%` is wired into `rules.neon`, `InArrayLooseComparisonRule` is opt-in with a false-positive fixture, and VPR-4 was implemented via a collector rather than deferred.

Three findings below, each with a reproduction.

---

## 1. `php vendor/bin/phpstan analyse` reports 4 errors; it was clean before

At 045225e this was `[OK] No errors`. At 1a66c78:

```
src/voku/PHPStan/Rules/IfConditionDiagnosticPolicy.php:106
src/voku/PHPStan/Rules/IfConditionDiagnosticPolicy.php:125
  Accessing PHPStan\Parser\LastConditionVisitor::ATTRIBUTE_NAME is not covered by
  backward compatibility promise. The class might change in a minor PHPStan version.
  [identifier=phpstanApi.classConstant]

src/voku/PHPStan/Rules/TraitContextComparisonCollector.php:79
  Identical: Possible insane comparison between null and PHPStan\Reflection\ClassReflection.
  [identifier=voku.Identical]
src/voku/PHPStan/Rules/TraitContextComparisonCollector.php:79
  Strict comparison using === between PHPStan\Reflection\ClassReflection and null
  will always evaluate to false. [identifier=identical.alwaysFalse]
```

Reproduce:

```bash
git checkout 1a66c78 && composer install
php vendor/bin/phpstan analyse
```

**`phpstanApi.classConstant`** — `LastConditionVisitor::ATTRIBUTE_NAME` sits outside PHPStan's BC promise, so referencing it fails this package's own analysis. One way out is to mirror the value (`'isLastCondition'`) as a local constant and pin it with a test asserting it still equals PHPStan's, so an upstream rename becomes a failing test rather than a check that silently stops firing. The test can reference the class freely, since `phpstan.neon` only analyses `src/` and `tests/fixtures/special/`.

**`TraitContextComparisonCollector.php:79`** — dead null-guard:

```php
$traitReflection = $scope->getTraitReflection();
if ($traitReflection === null) {   // unreachable: isInTrait() already returned true
    return null;
}
```

Not harmful, but it breaks the clean baseline — and it is this package's own rule that flags it, which is a decent dogfooding signal.

CI does not catch either, because the `Run phpstan` step is `continue-on-error: true` and only runs on PHP 7.4.

---

## 2. `reportDuplicateNativeComparisons: false` removes ~2 of 13 duplicate messages

The policy is wired into `IfConditionRule` and `TraitContextComparisonCollector` only. `IfConditionTernaryOperatorRule`, `IfConditionMatchRule`, `IfConditionSwitchCaseRule` and the boolean rules also process the same comparison node, and they still emit their copies.

Fixture:

```php
<?php
declare(strict_types=1);
namespace voku\PHPStan\Rules\Test\fixtures;

final class Dup
{
    public function inTernary(int $v): string
    {
        return $v == '' ? 'a' : 'b';          // line 9
    }

    public function inSwitch(int $v): string
    {
        switch (true) {
            case $v == '':                    // line 15
                return 'a';
            default:
                return 'b';
        }
    }
}
```

```neon
includes:
    - rules.neon
parameters:
    level: 9
    voku:
        reportDuplicateNativeComparisons: false
    paths:
        - /tmp/dup
```

```bash
php vendor/bin/phpstan analyse -c /tmp/dup.neon --autoload-file=vendor/autoload.php
```

PHPStan itself reports `equal.alwaysFalse` on both lines, so both are inside the intended overlap. Counting only `voku.*` messages:

| `reportDuplicateNativeComparisons` | messages |
| --- | --- |
| `true` | 13 |
| `false` | 11 |

What survives on line 9 with the flag off:

```
Equal:   Please do not use empty-string check for numeric values. …
Equal:   Possible insane comparison between '' and int.
Equal:   string ('') in combination with non-string (int) is not allowed.   (ExtendedBinaryOpRule)
Ternary: Condition between '' and int are falsy, please do not mix types.
Ternary: Please do not use empty-string check for numeric values. …
Ternary: Possible insane comparison between '' and int.
```

The three `Ternary:` lines restate what was just suppressed on `IfConditionRule`. The VPR-3 goal in #70 — "so that a project at level 4+ does not get the same finding three times" — is therefore largely unmet: exactly one message per line is removed.

Routing the same `IfConditionDiagnosticPolicy` through the ternary/match/switch/boolean rules brings the same fixture down to 2 remaining `voku.*` messages (`ExtendedBinaryOpRule`, which is a different concern and arguably should stay).

---

## 3. The trait deferral does not cover `ExtendedBinaryOpRule`

`TraitContextComparisonCollector` wraps `IfConditionRuleDiagnostics` only, so `ExtendedBinaryOpRule` / `ExtendedAssignOpRule` still report once per using class — the exact duplication VPR-4 measured.

Reproduce against the fixtures already in the repo:

```neon
includes:
    - rules.neon
parameters:
    level: 9
    paths:
        - tests/fixtures/TraitContext
```

```bash
php vendor/bin/phpstan analyse -c /tmp/trait.neon --autoload-file=vendor/autoload.php
```

```
LooseComparisonTrait.php (in context of class …\IntTraitConsumerOne):11:
  NotEqual: string ('') in combination with non-string (int) is not allowed.
LooseComparisonTrait.php (in context of class …\IntTraitConsumerTwo):11:
  NotEqual: string ('') in combination with non-string (int) is not allowed.
```

Two equivalent consumers, one trait line, two identical messages. The comparison diagnostics are correctly deduplicated; this one is not. Possibly in scope for #74.

---

## Supported PHPStan range (unchanged, still worth a decision)

`composer.json` declares `~2.0`, but the suite has never passed at the bottom of that range — this predates #70 and is not a regression:

| phpstan | result at 1a66c78 |
| --- | --- |
| 2.0.0 | 7 failures |
| 2.1.0 | 5 failures |
| 2.1.56 | green |
| 2.2.9 | green |

On 2.1.0 the 5 are the long-standing fixture tests (`IfConditionRuleTest` ×2, `IfConditionMatchRuleTest`, `IfConditionSwitchCaseRuleTest`, `IfConditionTernaryOperatorRuleTest`) whose expected message lists track a newer PHPStan's inference; the same 5 fail at 045225e. On 2.0.0 there are 2 more, both in `ComparisonDiagnosticPolicyTest`, which hard-codes 2.2's inference — deriving PHPStan's side at runtime (assert that the lines the flag removes equal the lines `ConstantLooseComparisonRule` reports) keeps that test valid on every 2.x.

Either fix those expectations for the whole range or raise the floor; right now the declared range is not tested and not true.

---

_Reviewed with the repository's own tooling: PHPUnit and `phpstan analyse` per version, using an isolated `vendor/` per checkout — a shared/symlinked `vendor/` resolves `voku\PHPStan\Rules\` to an absolute path and will silently analyse the wrong `src/`._

**Comment by voku** [comment]
Implemented and merged the #79 findings in bounded slices:

- #80 (`0f49d657`): restored clean self-analysis, removed the PHPStan API-excluded class-constant dependency, removed the dead trait guard, propagated duplicate-native comparison policy through ternary/match/switch/boolean paths, and made self-analysis blocking on PHP 8.4.
- #81 (`53b0c54e`): routed `ExtendedBinaryOpRule` trait diagnostics through the existing per-using-class collector boundary; equivalent contexts publish once and divergent contexts suppress context-specific findings.
- #82 (`1e469a07`): reproduced and fixed the same boundary gap for `ExtendedAssignOpRule` with a dedicated assign-op collector and end-to-end trait-context tests.
- #84 (`fe09a3c9`): raised the declared PHPStan floor to `^2.1.56` and pinned the PHP 7.4 CI lane to exactly `2.1.56`. The completed floor run explicitly installed PHPStan 2.1.56 and passed 130 tests / 703 assertions; PHP 8.4 current-PHPStan self-analysis also passed.

#83 was closed unmerged as superseded because its CI started before #82 merged and therefore did not exercise the final tree.

The `vendor/` symlink trap from this issue remains a reproduction caveat rather than a repository defect; CI uses isolated clean checkouts. No reconciliation of the old local `24770e0` branch is needed.
